src/lib/MerkleTree.ts

```diff
@@ -0,0 +1,246 @@
+import Cryptography from './Cryptography';
+
+/**
+ * Class containing a Merkle tree structure and its related operations.
+ */
+export default class MerkleTree {
+
+  /**
+   * Stores the Merkle tree root node once finalize() is called.
+   */
+  private merkleTreeRootNode?: MerkleNode;
+
+  /**
+   * Hashing function usinged by this Merkle tree.
+   */
+  private hash: (value: Buffer) => Buffer = Cryptography.sha256hash;
+
+  /**
+   * Intermediate structure for storing different sizes of balanced Merkle trees that will eventually form one final tree.
+   * [0] stores tree of 1 leaf, [1] stores tree of 2 leaves, [2] -> 4 leaves, [3] -> 8 leaves and so on.
+   */
+  private subtrees: (MerkleNode | undefined) [] = [];
+  private valueToMerkleNodeMap = new Map<Buffer, MerkleNode>();
```

Comment:
In sidetree's context why do we need this map, as we will be storing the values in IPFS. If we need it for linearization, then shouldn't that be taken care by the consumer of this class?

Reply:
The map is for quick construction of receipts. Linearization/serialization can be arguably implemented outside of this class, if and only if we require all values to be given in one shot, since the order of values is tightly couple with how the Merkle tree is constructed, so it makes sense to have the MerkleTree class dictate the order and the consumer can just call add(). But you brought up a good point that perhaps we require that all values needs to be given at the same time for now. So I will hide add(...) as a private function for now.